gtsam/basis/Fourier.h

```diff
@@ -0,0 +1,69 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file Fourier.h
+ * @brief Fourier decomposition, see e.g.
+ *        http://mathworld.wolfram.com/FourierSeries.html
+ * @author Varun Agrawal, Frank Dellaert
+ * @date July 4, 2020
+ */
+
+#pragma once
+
+#include <gtsam/basis/Basis.h>
+
+namespace gtsam {
+
+/// Fourier basis
+class GTSAM_EXPORT FourierBasis : public Basis<FourierBasis> {
+ public:
+  using Parameters = Eigen::Matrix<double, /*Nx1*/ -1, 1>;
+  using DiffMatrix = Eigen::Matrix<double, /*NxN*/ -1, -1>;
+
+  /**
+   *  Evaluate Real Fourier Weights of size N
+   *  e.g. N=5 yields bases: 1, cos(x), sin(x), cos(2*x), sin(2*x)
+   */
+  static Weights CalculateWeights(size_t N, double x) {
```

Comment:
CRTP args `a`/`b`

Reply:
Yeah this is an example of not making sense. A Fourier series is not defined in a certain interval (unlike a Chebyshev series which is explicitly defined on [-1,1].

Reply:
I'd argue it could still make sense - I may want it periodic with period 1 or with period 10.  So although I guess you wouldn't need both `a` and `b` (unless you're concerned about numerical precision of pi and doing calculations centered at e.g. 1e9) it could still make sense to define the period.

Reply:
You are correct. I propose throwing an exception below as not implemented yet. If someone needs it, they can implement. And, @gchenfc , if you feel called, just add a Pr after this is landed. It needs new unit tests and scaling the derivative weights, I think. It could be *very simple* as I think it is just multiplying the derivative weights with a scalar.

Reply:
Who is correct? :sweat_smile:

Reply:
I added something but I am going to leave the rest of this to @gchenfc or whomever needs it in the future. Not on my A->B but will assist if someone raises an issue.